dataprofiler/data_readers/data_utils.py

```diff
@@ -267,11 +271,102 @@ def read_json(
+    :type sample_nrows: int
+    :return: sampled values
+    :rtype: list
+    """
+    # Copyright 2021 Oscar Benjamin
+    #
+    # Permission is hereby granted, free of charge, to any person obtaining a copy
+    # of this software and associated documentation files (the "Software"), to deal
+    # in the Software without restriction, including without limitation the rights
+    # to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    # copies of the Software, and to permit persons to whom the Software is
+    # furnished to do so, subject to the following conditions:
+    #
+    # The above copyright notice and this permission notice shall be included in
+    # all copies or substantial portions of the Software.
+    #
+    # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    # IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    # FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    # AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    # SOFTWARE.
+    iterator = iter(file)
+    values = list(islice(iterator, sample_nrows))
+
+    irange = range(len(values))
+    indices = dict(zip(irange, irange))
+
+    kinv = 1 / sample_nrows
+    W = 1.0
+    while True:
+        W *= random() ** kinv
+        # random() < 1.0 but random() ** kinv might not be
+        # W == 1.0 implies "infinite" skips
+        if W == 1.0:
+            break
+        # skip is geometrically distributed with parameter W
+        skip = floor( log(random())/log1p(-W) )
+        try:
+            newval = next(islice(iterator, skip, skip+1))
+        except StopIteration:
+            break
+        # Append new, replace old with dummy, and keep track of order
+        remove_index = randrange(k)
+        values[indices[remove_index]] = None
+        indices[remove_index] = len(values)
+        values.append(newval)
+
+    values = [values[indices[i]] for i in irange]
+    _shuffle(values)
+    return values
+
+
+def rsample(
+        file_path: Union[str, TextIOWrapper],
+        args: dict
+) -> pd.DataFrame:
+    """
+    Implement Reservoir Sampling to sample n rows out of a total of M rows in the csv file.
+    This allows sampling without loading the entire file in memory.
+    :param file_path: path of the csv file to be read in
+    :type file: TextIOWrapper or string
+    :param args: options to read the csv file
+    :type args: dict
+    """
+
+    headers = []
+    header = args['header']
+    if not header:
+        result = reservoir(file_path, sample_nrows)
+    else:
+        if type(header) == list:
+            sort_header = sorted(header)
+        for i in range(header):
+        headers.append(next(file_path))
```

Comment:
not sure if gh having issues, but this looks like it needs a tab

Reply:
Do we need all of the prev info from the header or can we just grab the last line?